File: shouty/src/test/java/shouty/NetworkTest.java
package shouty;

import org.junit.Test;

import java.util.Arrays;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class NetworkTest {
    private int range = 100;
    private Network network = new Network(range);
    private String message = "Free bagels!";

    @Test
    public void broadcasts_a_message_to_a_listener_within_range() {
        Person sean = new Person("Sean", network, 0);
        Person lucy = mock(Person.class);
        network.subscribe(lucy);
        network.broadcast(message, sean);

        verify(lucy).hear(message);
    }

    @Test
    public void does_not_broadcast_a_message_to_a_listener_out_of_range() {
        Person sean = new Person("Sean:", network, 0);
        Person laura = mock(Person.class);
        when(laura.getLocation()).thenReturn(101);
        network.subscribe(laura);
        network.broadcast(message, sean);

        verify(laura, never()).hear(message);
    }

    @Test
    public void does_not_broadcast_a_message_to_a_listener_out_of_range_negative_distance() {
        Person sally = new Person("Sally", network, 101);
        Person lionel = mock(Person.class);
        when(lionel.getLocation()).thenReturn(0);
        network.subscribe(lionel);
        network.broadcast(message, sally);

        verify(lionel, never()).hear(message);
    }

    @Test
    public void does_not_broadcast_a_message_over_180_characters_even_if_listener_is_in_range() {
        Person sean = new Person("Sean", network, 0);

        char[] chars = new char[181];
        Arrays.fill(chars, 'x');
        String longMessage = String.valueOf(chars);

        Person laura = mock(Person.class);
        network.subscribe(laura);
        network.broadcast(longMessage, sean);

        verify(laura, never()).hear(longMessage);
    }

    @Test
    public void deducts_2_credits_for_a_shout_over_180_characters() {
        char[] chars = new char[181];
        Arrays.fill(chars, 'x');
        String longMessage = String.valueOf(chars);

        Person sean = new Person("Sean", network, 0, 0);
        Person laura = new Person("Laura", network, 0, 10);


        network.subscribe(laura);
        network.broadcast(longMessage, sean);

        assertEquals(-2, sean.credits);
    }

    @Test
    public void deducts_5_credits_for_mentioning_the_word_buy() {
        String message = "Come buy these awesome croissants";

        Person sean = new Person("Sean", network, 0, 100);
        Person laura = new Person("Laura", network, 0, 10);

        network.subscribe(laura);
        network.broadcast(message, sean);

        assertEquals(95, sean.credits);
    }

    @Test
    public void deducts_5_credits_for_mentioning_the_word_buy_several_times() {
        String message = "Come buy buy buy these awesome croissants";

        Person sean = new Person("Sean", network, 0, 100);
        Person laura = new Person("Laura", network, 0, 10);

        network.subscribe(laura);
        network.broadcast(message, sean);

        assertEquals(95, sean.credits);
    }

    @Test
    public void deducts_5_credits_if_the_word_buy_is_capitalized() {
        String message = "Come Buy these awesome croissants";

        Person sean = new Person("Sean", network, 0, 100);
        Person laura = new Person("Laura", network, 0, 10);

        network.subscribe(laura);
        network.broadcast(message, sean);

        assertEquals(95, sean.credits);
    }
}

File: content/bdd-with-cucumber/10/code/js/09-remove-todo-tag-and-bug-from-scenario/test/network.test.js
const assert = require("assert")
const { assertThat, equalTo } = require("hamjest")
const sinon = require("sinon")
const { Person, Network } = require("../src/shouty")

describe("Network", function () {
  const range = 100
  const network = new Network(range)
  const message = "Free bagels!"
  const longMessage = "x".repeat(181)

  it("broadcasts a message to a listener within range", function () {
    const shouterLocation = 0
    const listenerLocation = 90
    const sean = new Person("Sean", network, shouterLocation)
    const lucy = new Person("Lucy", network, listenerLocation)
    const lucyStub = sinon.spy(lucy)

    network.subscribe(lucy)
    network.broadcast(message, sean)

    assert.strictEqual(lucyStub.hear.getCall(0).args[0], message)
  })

  it("does not broadcast a message to a listener out of range", function () {
    const shouterLocation = 0
    const listenerLocation = 150
    const sean = new Person("Sean", network, shouterLocation)
    const lucy = new Person("Lucy", network, listenerLocation)
    const lucyStub = sinon.spy(lucy)

    network.subscribe(lucy)
    network.broadcast(message, sean)

    assert(lucyStub.hear.notCalled)
  })

  it("does not broadcast a message to a listener out of range negative distance", function () {
    const shouterLocation = 0
    const listenerLocation = -150
    const sean = new Person("Sean", network, shouterLocation)
    const lucy = new Person("Lucy", network, listenerLocation)
    const lucyStub = sinon.spy(lucy)

    network.subscribe(lucy)
    network.broadcast(message, sean)

    assert(lucyStub.hear.notCalled)
  })

  it("does not broadcast a message over 180 characters even if listener is in range", function () {
    const shouterLocation = 0
    const listenerLocation = 90
    const sean = new Person("Sean", network, shouterLocation)
    const lucy = new Person("Lucy", network, listenerLocation)
    const lucyStub = sinon.spy(lucy)

    network.subscribe(lucy)
    network.broadcast(longMessage, sean)

    assert(lucyStub.hear.notCalled)
  })

  it("can change the range", function () {
    const sean = new Person("Sean", network, 0)
    const laura = new Person("Laura", network, 10)

    network.broadcast(message, sean)
    assert.deepEqual(laura.messagesHeard(), [message])

    network.range = 5
    network.broadcast(message, sean)
    assert.deepEqual(laura.messagesHeard(), [message])
  })

  context("credits", () => {
    it("deducts 2 credits for a shout over 180 characters", () => {
      const longMessage = "x".repeat(181)

      const sean = new Person("Sean", network, 0)
      const laura = new Person("Laura", network, 10)

      network.subscribe(laura)
      network.broadcast(longMessage, sean)

      assertThat(sean.credits, equalTo(-2))
    })

    it("deducts 5 credits for mentioning the word 'buy'", () => {
      const message = 'Come buy these awesome croissants'

      const sean  = new Person("Sean", network, 0, 100)
      const laura = new Person("Laura", network, 10)

      network.subscribe(laura)
      network.broadcast(message, sean)

      assertThat(sean.credits, equalTo(95))
    })

    it("deducts 5 credits even if you mention the word 'buy' several times", () => {
      const message = 'Come buy buy buy these awesome croissants'

      const sean  = new Person("Sean", network, 0, 100)
      const laura = new Person("Laura", network, 10)

      network.subscribe(laura)
      network.broadcast(message, sean)

      assertThat(sean.credits, equalTo(95))
    })
  })
})
